File: Server/src/plugin/interaction/item/PharoahSceptre.java
package plugin.interaction.item;

import org.crandor.cache.def.impl.ItemDefinition;
import org.crandor.game.content.dialogue.DialoguePlugin;
import org.crandor.game.interaction.OptionHandler;
import org.crandor.game.node.Node;
import org.crandor.game.node.entity.player.Player;
import org.crandor.game.node.entity.player.info.portal.Perks;
import org.crandor.game.node.item.Item;
import org.crandor.game.system.task.Pulse;
import org.crandor.game.world.GameWorld;
import org.crandor.game.world.map.Location;
import org.crandor.game.world.update.flag.context.Animation;
import org.crandor.game.world.update.flag.context.Graphics;
import org.crandor.plugin.InitializablePlugin;
import org.crandor.plugin.Plugin;


/**
 * Adds functionality to the pharoah's scepter
 * @author ceik
 */
@InitializablePlugin
public final class PharoahSceptre extends OptionHandler {
    private static final Graphics GRAPHICS = new Graphics(308, 100, 50);
    private static final Animation ANIMATION = new Animation(714);

    @Override
    public Plugin<Object> newInstance(Object arg) throws Throwable {
        new SceptreDialog().init();
        ItemDefinition.forId(9044).getConfigurations().put("option:teleport",this);
        ItemDefinition.forId(9046).getConfigurations().put("option:teleport",this);
        ItemDefinition.forId(9048).getConfigurations().put("option:teleport",this);
        ItemDefinition.forId(9050).getConfigurations().put("option:teleport",this);
        return this;
    }
    @Override
    public boolean handle(Player player, Node node, String string){
        Item item = node.asItem();
        if(string.equals("teleport")){
            switch(item.getId()) {
                case 9044:
                case 9046:
                case 9048:
                    player.getDialogueInterpreter().open(325128389);
                    return true;
                case 9050:
                    player.getPacketDispatch().sendMessage("You have used up all the charges on this sceptre.");
                    return true;
            }
        }
        return true;
    }

    public final class SceptreDialog extends DialoguePlugin {
        public SceptreDialog() {
            /**
             * empty
             */
        }
        public SceptreDialog(Player player){super(player);}

        @Override
        public DialoguePlugin newInstance(Player player){ return new SceptreDialog(player);}

        @Override
        public boolean open(Object... args){
            //Item item = (Item) args[0];
            player.getPacketDispatch().sendMessage("" + args[0]);
            interpreter.sendOptions("Teleport to:","Jalsavrah","Jaleustrophos","Jaldraocht","Nowhere");
            return true;
        }
        public void teleport(Location location, Player player){
            player.lock();
            player.visualize(ANIMATION, GRAPHICS);
            player.getImpactHandler().setDisabledTicks(4);
            GameWorld.submit(new Pulse(4, player) {
                @Override
                public boolean pulse() {
                    player.unlock();
                    player.getProperties().setTeleportLocation(location);
                    player.getAnimator().reset();
                    return true;
                }
            });
        }
        @Override
        public boolean handle(int interfaceId, int buttonId){
            if (player.isTeleBlocked()) {
                player.sendMessage("A magical force has stopped you from teleporting.");
                return false;
            }
            switch(buttonId){
                case 1:
                    teleport(Location.create(3289,2786,0), player);
                    stage = 10;
                    break;
                case 2:
                    teleport(Location.create(3342,2827,0),player);
                    stage = 10;
                    break;
                case 3:
                    teleport(Location.create(3233,2902,0),player);
                    stage = 10;
                    break;
                case 4:
                    stage = 20;
                    break;
            }
          switch(stage){
              case 10:
                  if(player.getInventory().containsItem(new Item(9044))){
                      player.getInventory().remove(new Item(9044));
                      player.getInventory().add(new Item(9046));
                      player.getPacketDispatch().sendMessage("<col=7f03ff>Your Pharoah's Sceptre has 2 charges remaining.");
                      end();
                  } else if (player.getInventory().containsItem(new Item(9046))){
                      player.getInventory().remove(new Item(9046));
                      player.getInventory().add(new Item(9048));
                      player.getPacketDispatch().sendMessage("<col=7f03ff>Your Pharoah's Sceptre has 1 charge remaining.");
                      end();
                  } else if (player.getInventory().containsItem(new Item(9048))){
                      player.getInventory().remove(new Item(9048));
                      player.getInventory().add(new Item(9050));
                      player.sendMessage("<col=7f03ff>Your Pharoah's Sceptre has used its last charge.");
                      end();
                  }
                  stage = 20;
                  break;
              case 20:
                  end();
                  break;
          }
            return true;
        }
        public int[] getIds() { return new int[] {325128389};}
    }

}

File: Server/src/main/java/core/game/interaction/item/PharoahSceptre.kt
package core.game.interaction.item

import api.EquipmentSlot
import api.openDialogue
import api.sendMessage
import rs09.game.world.GameWorld.Pulser
import rs09.game.content.activity.pyramidplunder.PyramidPlunderMinigame.Companion.GUARDIAN_ROOM
import core.plugin.Initializable
import core.game.node.entity.player.Player
import core.game.node.item.Item
import core.game.system.task.Pulse
import core.game.world.map.Location
import core.game.world.update.flag.context.Animation
import core.game.world.update.flag.context.Graphics
import org.rs09.consts.Items
import rs09.game.content.dialogue.DialogueFile
import rs09.game.interaction.InteractionListener
import rs09.tools.END_DIALOGUE

/**
 * Adds functionality to the pharoah's scepter
 * @author ceik
 */
class PharoahSceptre : InteractionListener() {
    override fun defineListeners() {
        val SCEPTRES = intArrayOf(Items.PHARAOHS_SCEPTRE_9044, Items.PHARAOHS_SCEPTRE_9046, Items.PHARAOHS_SCEPTRE_9048, Items.PHARAOHS_SCEPTRE_9050)

        on(SCEPTRES, ITEM, "teleport", "operate"){player, node ->
            val sceptre = node.asItem()

            if(sceptre.id == SCEPTRES.last())
            {
                sendMessage(player, "You have used up all the charges on this sceptre.")
                return@on true
            }

            openDialogue(player, SceptreDialog())
            return@on true
        }
    }

    class SceptreDialog : DialogueFile() {
        fun teleport(location: Location?, player: Player) {
            player.lock()
            player.visualize(ANIMATION, GRAPHICS)
            player.impactHandler.disabledTicks = 4
            Pulser.submit(object : Pulse(4, player) {
                override fun pulse(): Boolean {
                    player.unlock()
                    player.properties.teleportLocation = location
                    player.animator.reset()
                    return true
                }
            })
        }

        override fun handle(componentID: Int, buttonID: Int) {
            if (player!!.isTeleBlocked) {
                player!!.sendMessage("A magical force has stopped you from teleporting.")
                return
            }
            when (stage) {
                0 -> options("Jalsavrah", "Jaleustrophos", "Jaldraocht", "Nowhere").also { stage++ }
                1 -> {
                    end()
                    when (buttonID) {
                        1 -> {
                            teleport(GUARDIAN_ROOM, player!!)
                        }
                        2 -> {
                            teleport(Location.create(3342, 2827, 0), player!!)
                        }
                        3 -> {
                            teleport(Location.create(3233, 2902, 0), player!!)
                        }
                        4 -> return
                    }
                    //This sucks but I'm too lazy to fix it.
                    if (player!!.equipment.containsItem(Item(9044))) {
                        player!!.equipment.replace(Item(9046), EquipmentSlot.WEAPON.ordinal)
                        player!!.packetDispatch.sendMessage("<col=7f03ff>Your Pharoah's Sceptre has 2 charges remaining.")
                    } else if (player!!.equipment.containsItem(Item(9046))) {
                        player!!.equipment.replace(Item(9048), EquipmentSlot.WEAPON.ordinal)
                        player!!.packetDispatch.sendMessage("<col=7f03ff>Your Pharoah's Sceptre has 1 charge remaining.")
                    } else if (player!!.equipment.containsItem(Item(9048))) {
                        player!!.equipment.replace(Item(9050), EquipmentSlot.WEAPON.ordinal)
                        player!!.sendMessage("<col=7f03ff>Your Pharoah's Sceptre has used its last charge.")
                    } else if (player!!.inventory.containsItem(Item(9050))) {
                        player!!.inventory.remove(Item(9050))
                        player!!.inventory.add(Item(9048))
                        player!!.packetDispatch.sendMessage("<col=7f03ff>Your Pharoah's Sceptre has 2 charges remaining.")
                    } else if (player!!.inventory.containsItem(Item(9048))) {
                        player!!.inventory.remove(Item(9048))
                        player!!.inventory.add(Item(9046))
                        player!!.packetDispatch.sendMessage("<col=7f03ff>Your Pharoah's Sceptre has 1 charge remaining.")
                    } else if (player!!.inventory.containsItem(Item(9046))) {
                        player!!.inventory.remove(Item(9046))
                        player!!.inventory.add(Item(9044))
                        player!!.sendMessage("<col=7f03ff>Your Pharoah's Sceptre has used its last charge.")
                    }
                }
            }
        }
    }

    companion object {
        private val GRAPHICS = Graphics(715)
        private val ANIMATION = Animation(714)
    }
}